### PR TITLE
devel: Add the design document

### PR DESCRIPTION
**Pull request by EdDev**
Port the design document from google-docs and make it available for
upstream.

### CODE REVIEW

File: devel/design/networking-api.md [hunk 1]
```diff
@@ -0,0 +1,509 @@
...
+           +--ro admin-status
+           +---ro link-status (up|down)
+           +---ro link-speed (base link speed for virtual interfaces)
+           +-- ethernet:
+              +-- rw auto-negotiation (yes|no)
+              +-- rw duplex (ro with auto-negotiation)
+              +-- rw speed (ro with auto-negotiation)
+              +-- rw flow-control (ro with auto-negotiation)
+           +--rw mac-address
+           +--rw vlan-id (for vlan type)
```
Comment (EdDev) [inline devel/design/networking-api.md]: In order to be consistent with the other type specific interfaces, these two properties (`vlan-id` and `vlan-base`) should be placed under a `vlan` subtree.
```
+--vlan:
   +-- id
   +-- base-iface

File: devel/design/networking-api.md [hunk 2]
```diff
@@ -0,0 +1,509 @@
+# Linux Networking API
+
+## OVERVIEW
+Linux networking has always been a major subsystem in the Linux ecosystem. Its functionality is used broadly across domains, from connecting a local host to the network up to a fully functional network device (router, switch, traffic shaper, etc).
```
Comment (tyll) [inline devel/design/networking-api.md]: It would be better to wrap the long lines.

**Review by tyll — APPROVED** [review]

File: devel/design/networking-api.md (hunk 2 above)
Comment (dankenigsberg) [reply devel/design/networking-api.md]: Linux networking has -> Networking has

(Avoid repetition)

File: devel/design/networking-api.md (hunk 1 above)
Comment (EdDev) [reply devel/design/networking-api.md]: Done

File: devel/design/networking-api.md (hunk 2 above)
Comment (EdDev) [reply devel/design/networking-api.md]: Done (both)

**Comment by cathay4t** [comment]
This document contains too much information:
 * Propaganda
 * Existing projects trying to solve similar problem and their pros/cons.
 * The nmstatectl examples.
 * Tech details not related to this project -- NETCONF(there is no example or
   support metrix about nmstate in netconf client)

My concerns:
 * Who is the reader of this document?
    * Potential developer who want to use libnmstate in their project.
      They don't care about propaganda and nmstate internal design choices,
      but need document about how to use it, what nmstate supports and detailed
      document on every possible value of outputs.
    * Potential devloper who want to help nmstate coding.
      They need nmstate internal workflow, code style, test layout, and
      who/where to ask if need help.
    * IT admin who want to use `nmstatectl`.

I would like to see neat documents targeted to different user/developer in case they
say TLNR.

**Comment by EdDev** [comment]
@cathay4t  , I do not think this document includes [propaganda](https://en.wikipedia.org/wiki/Propaganda).
This is NMState high level design document, pre-dating the implementation (by several months). Its purpose includes reasoning for the need of such a solution,  present the existing alternatives considered at the time, giving enough background to persuade the audience why this solution is worth an implementation.
It includes details on what it needs to solve and at a high level, how to do it.
It embeds feedback gathered from reviewers and enough information to define NMState domain.

I would expect any engineer involved with this project to read it in order to understand the big picture.
The document is linked from the `devel` section, not aimed for NMState users directly, but it is informative for users who want to understand its source of existence.

**Comment by tyll** [comment]
I can understand @cathay4t's point of too much information in the document, especially since there is no table of contents. Therefore it might make sense to split the document into multiple files and then have a TOC that links to the individual parts.

**Comment by EdDev** [comment]
I understand that it can be improved, but the point was to publicize the in-house design document, not to re-write it. I am more than happy to accept improvements to it after it gets in.

If there are specific changes that are easy to edit now, I am good with changing it. Just point to the specific issues. If there are big changes, please suggest them as follow-up PR/s.

**Comment by tyll** [comment]
> I understand that it can be improved, but the point was to publicize the in-house design document, not to re-write it. I am more than happy to accept improvements to it after it gets in.

works for me

**Review by cathay4t — APPROVED** [review]

**Comment by cathay4t** [comment]
Merged.

**Pull request merged**